runtime/src/accounts_index.rs

```diff
@@ -376,6 +412,43 @@ impl<T: 'static + Clone> AccountsIndex<T> {
         }
     }
 
+    fn do_scan_mint_accounts<F>(
+        &self,
+        ancestors: &Ancestors,
+        mut func: F,
+        mint_key: &Pubkey,
+        max_root: Option<Slot>,
+    ) where
+        F: FnMut(&Pubkey, (&T, Slot)),
+    {
+        let mint_keys: Vec<Pubkey> = {
+            let mint_keys = self.mint_index.index.get(mint_key);
+            if mint_keys.is_none() {
+                return;
+            }
+            let mint_keys = mint_keys.unwrap();
+            if mint_keys.is_empty() {
+                return;
+            }
+            mint_keys
+                .value()
+                .iter()
+                .map(|pubkey_ref| *pubkey_ref)
+                .collect()
+        };
+
+        for pubkey in mint_keys {
+            // Maybe these reads from the AccountsIndex can be batched everytime it
+            // grabs the read lock as well...
+            if let Some((list_r, index)) = self.get(&pubkey, Some(ancestors), max_root) {
+                func(
```

Comment:
I guess we need similar check as the original code had there: https://github.com/solana-labs/solana/pull/14212/files?diff=split&w=1#r546409102.

The reason is that mint_index could be stale (maybe?) if the account is closed and used for accounts which isn't owned by mint.

Reply:
Yup that's a good catch! if you have a pubkey with a mint in slot 0, and then it's updated to a non-owned account in slot 1 in the same fork, then the initial code would still return the slot 0 entry even when there's a scan done on slot 1. The original check has been re-added and @CriesofCarrots has added a test here as well: https://github.com/solana-labs/solana/pull/14212/files#diff-ed47b4a0198313377e091bb3957bbbc63d937805426d1b2b6de39d0a50d32a0cR8667